Add OmniSharp ExternalAccess

333fred opened the pull request:
We had a discussion some months ago about adding an ExternalAccess layer for OmniSharp, as they use a decent amount of private reflection for various services. I've gone through the O# codebase and looked for everything that is using reflection to work. This is the first time I've set up one of these layers, so please let me know what I need to change.

CyrusNajmabadi commented:
Was this modeled after the f# ea impl? Thanks!

333fred commented:
> 
> 
> Was this modeled after the f# ea impl? Thanks!

Yes. I can't say for certain it was modelled _correctly_, but the intent was to mirror how F# does it. From what I could see though, F# mostly plugs into existing Roslyn infrastructure (which is then used by VS), whereas O# does a bit of that, and a bit of calling infrastructure itself, serving as the VS side here.

CyrusNajmabadi commented:
I think a big part of the f# already here (if I understand it correctly) is that it's actually a separate dll for the ea layer. This differs from ts, where the ea layer is mixed into the main code for that layer.

333fred commented:
> I think a big part of the f# layer here (if I understand it correctly) is that it's actually a separate dll for the ea layer. This differs from ts, where the ea layer is mixed into the main code for that layer. 

The intention with my change here is that we have a dedicated MS.CA.EA.OS dll.

CyrusNajmabadi commented:
Great. Thanks for the clarification. I can review tomorrow once @sharwell Looks. Thanks!

333fred commented on src/Tools/ExternalAccess/OmniSharp/Internal/ExtractClass/OmniSharpExtractClassOptionsService.cs:
```diff
@@ -0,0 +1,39 @@
+namespace Microsoft.CodeAnalysis.ExternalAccess.OmniSharp.Internal.ExtractClass
+{
+    [Shared]
+    [ExportWorkspaceService(typeof(OmniSharpExtractClassOptionsService))]
```
```suggestion
    [ExportWorkspaceService(typeof(IExtractClassOptionsService))]
```

sharwell commented on src/Tools/ExternalAccess/OmniSharp/MetadataAsSource/OmniSharpMetadataAsSourceService.cs:
```diff
@@ -0,0 +1,30 @@
+        /// <param name="symbol">The symbol to generate source for</param>
+        /// <param name="cancellationToken">To cancel document operations</param>
+        /// <returns>The updated document</returns>
+        public static Task<Document> AddSourceToAsync(Document document, Compilation symbolCompilation, ISymbol symbol, CancellationToken cancellationToken = default)
```
💡 `CancellationToken` shouldn't be optional for non-public APIs

```suggestion
        public static Task<Document> AddSourceToAsync(Document document, Compilation symbolCompilation, ISymbol symbol, CancellationToken cancellationToken)
```

333fred replied:
Fair enough, but this was a copy of the existing signature :).

sharwell commented on src/Tools/ExternalAccess/OmniSharp/Internal/PickMembers/OmniSharpPickMembersService.cs:
```diff
@@ -0,0 +1,33 @@
+namespace Microsoft.CodeAnalysis.ExternalAccess.OmniSharp.Internal.PickMembers
+{
+    [Shared]
+    [ExportWorkspaceService(typeof(IPickMembersService))]
```
💭 This is a host-layer service right?

333fred replied:
Yes, the existing impl is `VisualStudioPickMembersService`. Should I be setting an explicit layer here?

jasonmalinowski replied:
I would have assumed this needs to be marked as "Host" for the service layer, or otherwise this and the other one are both "Default", and I'm not sure which MEF service is going to win there.

333fred replied:
So I _should_ do this, or I shouldn't? Happy to mark it as host, just need to know I should.

> I'm not sure which MEF service is going to win there.

I would never expect this service to be loaded in the same process as the VS host version, fwiw.

jasonmalinowski replied:
Oh, I see -- this is a service that *has* no default implementation -- the only implementation comes from the host. I was worried this was going to conflict with the default impl, but there is no such thing. So this will work, but that said, if you mark it Host it'll be marked the same way like everything else, which sure won't hurt anything. When in Rome...

333fred replied:
Done.

sharwell commented on src/Tools/ExternalAccess/OmniSharp/ImplementType/OmniSharpImplementTypeOptions.cs:
```diff
@@ -0,0 +1,31 @@
+    internal static class OmniSharpImplementTypeOptions
+    {
+        public static PerLanguageOption<OmniSharpImplementTypeInsertionBehavior> InsertionBehavior =
+            new PerLanguageOptionInternalConverter<OmniSharpImplementTypeInsertionBehavior, ImplementTypeInsertionBehavior>(ImplementTypeOptions.InsertionBehavior);
```
❔ Don't we have an `AsPublicOption` extension or something similar?

sharwell replied:
Since `ImplementTypeInsertionBehavior` isn't public, I would actually recommend using helper methods to get/set the option, and not actually exposing the option directly.

333fred commented:
@sharwell addressed feedback. I also created a separate `.CSharp` dll for OmniSharp, as they also have a layering system like we do where their `.Roslyn` project only references the common base, and `.Roslyn.CSharp` is where they add in the `.CSharp` roslyn dlls.

jasonmalinowski commented on src/Tools/ExternalAccess/OmniSharp/Microsoft.CodeAnalysis.ExternalAccess.OmniSharp.csproj:
```diff
@@ -0,0 +1,45 @@
+    <TargetFramework>netstandard2.0</TargetFramework>
+
+    <IsPackable>true</IsPackable>
+    <PackageId>Microsoft.CodeAnalysis.ExternalAccess.OmniSharp</PackageId>
```
Does this need to be marked as not going up to NuGet just like https://github.com/dotnet/roslyn/pull/53007?

dibarbet replied:
@jasonmalinowski still having a chat with those folks - I'm not 100% sure that is the solution as we've never had that marker yet we didn't use to upload those packages.

333fred replied:
Of note, all our current external access packages are on nuget: https://www.nuget.org/packages?q=Microsoft.CodeAnalysis.ExternalAccess

JoeRobich replied:
Sigh, for 5.0.200 there was a goof up and all the packages were automatically published.

jasonmalinowski replied:
@JoeRobich do we have a plan to unlist those? I know I saw some mail thread about unlisting due to a goof but not sure if it's more fallout of this or something else.

jasonmalinowski replied:
In any case, so what's the recommendation for @333fred here, is there anything that we know that _is_ correct and can be put in this PR, or he should merge and we have a follow-up that cleans up his stuff along with everything else?

JoeRobich replied:
Adding a property should not be necessary as we are responsible for publishing our own packages.

333fred replied:
Alright, marking as resolved then.

333fred commented:
I rebased this to squash the fixup commits. The two commits at the end, adding a unit test project to assert we're keeping enums in sync, are now.

333fred commented:
I think the main thing this is waiting for is a strong-name key for omnisharp. Their assemblies are unsigned today, which doesn't work with IVT as roslyn is signed. I'll work on getting them signed in the next couple of days and add the public key info here when I do.

333fred commented:
https://github.com/OmniSharp/omnisharp-roslyn/pull/2143 is now merged, and this PR is updated to add OmniSharp's signing key directly.

333fred commented:
@jasonmalinowski @JoeRobich you had a couple of questions I requested more info on: is there anything I need to do? Or is this ready for merge?

jasonmalinowski commented on src/Tools/ExternalAccess/OmniSharp/Structure/OmniSharpBlockTypes.cs:
```diff
@@ -0,0 +1,33 @@
+    internal static class OmniSharpBlockTypes
+    {
+        // Basic types.
+        public static string Nonstructural => BlockTypes.Nonstructural;
```
Do you want any sort of test that these match? Your enum test was nice but not sure it's easy to make it work for this too.

333fred replied:
I tried to think of a way to do it neatly, but it would just be a bunch of reflection over the properties and I didn't feel it was worth it.

jasonmalinowski commented on src/Tools/ExternalAccess/OmniSharp/Internal/CodeRefactorings/WorkspaceServices/OmniSharpSymbolRenamedCodeActionOperationFactoryWorkspaceService.cs:
```diff
@@ -0,0 +1,32 @@
+{
+    [Shared]
+    [ExportWorkspaceService(typeof(ISymbolRenamedCodeActionOperationFactoryWorkspaceService))]
+    class OmniSharpSymbolRenamedCodeActionOperationFactoryWorkspaceService : ISymbolRenamedCodeActionOperationFactoryWorkspaceService
```
How was Omnisharp using this? We have a PR that will ultimately remove this (https://github.com/dotnet/roslyn/pull/47579) and mostly curious.

333fred replied:
https://github.com/OmniSharp/omnisharp-roslyn/blob/ff21ac475c/src/OmniSharp.Roslyn/WorkspaceServices/OmniSharpSymbolRenamedCodeActionOperationFactoryWorkspaceService.cs

JoeRobich replied:
Specify access modifier?

jasonmalinowski reviewed (approved):
So generally OK with this but want two things:

1. Would like a signoff from @JoeRobich or @dibarbet that the NuGet upload handling is correct, or at least as correct as this PR can have it if it's still under investigation.
2. We get something written up in the docs folder explaining what the ownership process is for this.

For that second item specifically, what the breaking change policy is. For some of our existing external access projects, we can't make any breaking change until we've checked with the team to make sure it's not being used, or we do a complicated dance; the IDE team is well aware of that process since by now I think everybody has personally failed to do it at least once and learned the hard way. Since Omnisharp still chooses to upgrade to newer Roslyn versions (right?) I presume the process here is something more where if we're making a change we'd want to notify them (and we should clarify who "them" is and what communication medium is chosen) so at least they have a heads up that it's coming so they can adapt on next upgrade.

And the docs here can be a single sentence of "ping <these folks> on GitHub, wait until somebody gives you a thumbs up, and you're good". Just want something written down here.

333fred commented:
@jasonmalinowski added a doc.

jasonmalinowski commented:
@333fred looks great, thanks!
